lib/nrf_modem_lib/lte_connectivity/lte_connectivity.c

```diff
@@ -263,17 +385,40 @@ static void pdn_event_handler(uint8_t cid, enum pdn_event event, int reason)
 	}
 }
 
+static void lte_reg_handler(const struct lte_lc_evt *const evt)
+{
+	if (evt->type != LTE_LC_EVT_NW_REG_STATUS) {
+		return;
+	}
+
+	switch(evt->nw_reg_status) {
+		case LTE_LC_NW_REG_REGISTERED_HOME:
+			__fallthrough;
+		case LTE_LC_NW_REG_REGISTERED_ROAMING:
+			/* Cancel the registration timeout and mark serving cell as available. */
+			LOG_DBG("Registered to serving cell");
+			k_work_cancel_delayable(&reg_timeout_work);
+			update_has_cell(true);
+			break;
+		default:
+			LOG_DBG("Not registered to serving cell");
+			/* Start a timeout to mark the serving cell as lost. */
+			reg_timeout_schedule();
+			break;
```

Comment:
I'll discuss this approach more next week when few of my colleagues are back in the office.
Meanwhile, can you describe the problems you face with the PDN activated/deactivated events. I suspect they are coming too late but how much is that?
I assume your application is trying to send data and then connection is lost, which is informed late to the application via `PDN_EVENT_DEACTIVATED`.

Reply:
At least `+CEREG: 2` (`LTE_LC_NW_REG_SEARCHING`) should not be considered as loss of connectivity immediately, because modem may find a new (or the same) suitable cell quickly. Guidance from the modem team has been that the application side should rely on the PDN events to detect whether connectivity is available or not. I think that the modem most likely already has a similar timer which is being implemented here, but of course that's not configurable by the application.

I would also like to understand what is the actual problem with the PDN deactivated event?

Reply:
> can you describe the problems you face with the PDN activated/deactivated events. I suspect they are coming too late but how much is that?

We have been simulating network loss by placing our devices into RF chambers, and the PDN deactivated event never came, even after many hours of waiting.

My intuition for this is that the PDN status indicates whether there is a logical channel opened for the device, and the CEREG status represents whether there is a PHY layer channel opened for the device

Losing the cell does not immediately imply loss of the logical channel since if cell is regained in a "timely" fashion, the device may resume using the logical channel / PDN carrier it had been using before with no issue

So, in **_theory_**, yes, PDN events are the most "accurate" from the perspective of the carrier -- so long as a PDN is available, the carrier is allowing the device to connect to the internet.

However, from an application perspective, it may be desirable to consider the connection lost much sooner. An application may not want to wait for a non-configurable amount of time potentially lasting hours for the PDN carrier to be officially considered lost. An application may want to consider even a few minutes of loss-of-cell as a disconnection from the network.

These changes allow that behavior to be configured -- applications can rely solely on PDN if they want, or they can set up a time out to their specifications based on what would be considered to them an abnormal amount of time without a serving cell

Reply:
> At least +CEREG: 2 (LTE_LC_NW_REG_SEARCHING) should not be considered as loss of connectivity immediately

Regarding CEREG: 2 specifically: Good point, I think searching generally is pretty fast, so it would probably suffice to just have this state ignored completely by the timeout system. I will make that change in the coming days

Reply:
Can you provide us modem logs from the case where PDN deactivation is not seen or takes tens of minutes?
And did you see somewhere that the connection was really lost in an RF chamber. I've used one which still doesn't fully disconnect the device but it keeps connecting and disconnecting so the RF chamber is leaking. It would be good to see modem logs to know what's going on.

Reply:
Discussed the aforementioned logs offline, and yes PDN is definitely not coming in, connection is definitely being lost in the RF chamber, and all of this is expected behavior.

@tokangas I've adjusted the event logic to ignore CEREG 2, which I think should do the trick. I'll be testing with this patch over on #11519

Reply:
This comment can be closed as we've confirmed from modem team that we should be monitoring `CEREG: 4` to determine that network is unavailable.